### PR TITLE
Add disable push for proxy project

### PR DESCRIPTION
**Pull request by stonezdj**
Add middleware for blob and manifest push operation

Signed-off-by: stonezdj <stonezdj@gmail.com>

### CODE REVIEW

File: src/server/registry/route.go [hunk 1]
```diff
@@ -82,6 +84,7 @@ func RegisterRoutes() {
 	root.NewRoute().
 		Method(http.MethodPut).
 		Path("/*/blobs/uploads/:session_id").
+		Middleware(repoproxy.BlobPushMiddleware()).
```
Comment (ywk253100) [inline src/server/registry/route.go:87]: Block patch method as well

File: src/server/middleware/repoproxy/proxy.go [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+					errors.Errorf("can not push artifact to a proxy project: %v", p.Name)))
+			return
+		}
+		next.ServeHTTP(w, r)
+		return
+	})
+}
+
+// BlobPushMiddleware disable blob push for non proxy session
+func BlobPushMiddleware() func(http.Handler) http.Handler {
```
Comment (ywk253100) [inline src/server/middleware/repoproxy/proxy.go:77]: `ManifestPushMiddleware` and `BlobPushMiddleware`  can be merged into one?

File: src/server/middleware/repoproxy/proxy.go [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+}
+
+// isProxySession check if current security context is proxy session
+func isProxySession(ctx context.Context) bool {
+	sc, ok := security.FromContext(ctx)
+	if !ok {
+		log.Error("Failed to get security context")
+		return false
+	}
+	if sc.Name() == securitySecret.SecurityContextName &&
```
Comment (ywk253100) [inline src/server/middleware/repoproxy/proxy.go:48]: You can use `SecurityContext.IsSolutionUser()` instead to replace the logic `if sc.Name() == securitySecret.SecurityContextName`

File: src/server/middleware/repoproxy/proxy.go (hunk 2 above)
Comment (reasonerjt) [reply src/server/middleware/repoproxy/proxy.go:77]: +1

File: src/server/registry/route.go (hunk 1 above)
Comment (stonezdj) [reply src/server/registry/route.go:87]: updated

File: src/server/middleware/repoproxy/proxy.go (hunk 2 above)
Comment (stonezdj) [reply src/server/middleware/repoproxy/proxy.go:77]: Merged

File: src/server/middleware/repoproxy/proxy.go (hunk 3 above)
Comment (stonezdj) [reply src/server/middleware/repoproxy/proxy.go:48]: Updated

File: src/common/security/secret/context.go [hunk 4]
```diff
@@ -20,6 +20,9 @@ import (
 	"github.com/goharbor/harbor/src/pkg/permission/types"
 )
 
+// SecurityContextName ...
+const SecurityContextName = "secret"
```
Comment (ywk253100) [inline src/common/security/secret/context.go:24]: The changes for this file isn't needed?

File: src/server/middleware/repoproxy/proxy.go [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+		ctx := r.Context()
+		art := lib.GetArtifactInfo(ctx)
+		p, err := project.Ctl.GetByName(ctx, art.ProjectName)
+		if err != nil {
+			serverHttp.SendError(w, err)
+			return
+		}
+		if isProxyProject(p) && !isProxySession(ctx) {
+			serverHttp.SendError(w,
+				errors.ForbiddenError(
```
Comment (ywk253100) [inline src/server/middleware/repoproxy/proxy.go:64]: Is 403 a correct status code for this case? How about 405? @reasonerjt @stonezdj

Comment (stonezdj) [reply src/server/middleware/repoproxy/proxy.go]: Use 405

**Pull request merged**